src/os/bluestore/BlueStore.h

```diff
@@ -1554,16 +1556,26 @@ class BlueStore : public ObjectStore,
       q.push_back(*txc);
     }
 
-    void flush() override {
+    void drain() {
       std::unique_lock<std::mutex> l(qlock);
       while (!q.empty())
 	qcond.wait(l);
     }
 
-    void drain() {
+    void flush() override {
       std::unique_lock<std::mutex> l(qlock);
-      while (!q.empty())
+      while (true) {
+	if (q.empty()) {
+	  return;
+	}
+	TransContext *txc = &q.back();
```

Comment:
what's about making a func is_all_kv_submitted() that retrieves the last entry and check its state for KV_SUBMITTED. And start using it both here and prior to notification sending? This way one can have just a single indication at proper moment and limit corresponding mutex/thread switch overhead...